File: 144-BiTreePreOrderTrave/recursive.java
class Solution {
    private List<Integer> res;

    public List<Integer> preorderTraversal(TreeNode root) {
        res = new ArrayList<Integer>();
        if (root != null)
            helper(root);
        return res;
    }

    public void helper(TreeNode root) {
        if (root == null) return;
        res.add(root.val);
        helper(root.left);
        helper(root.right);
    }
}

public class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) { val = x; }
}

File: 124-BiTreeMaxPathSum/recursive.cpp
#include <iostream>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
private:
    int maxSum;
public:
    int maxPathSum(TreeNode* root) {
        maxSum = -2147483648;
        helper(root);
        return maxSum;
    }
    
    int helper(TreeNode* root) {
        int temp = 0, res = 0;
        if (root->left) {
            int left = helper(root->left);
            if (left > 0) temp += left;
            res = max(res, left);
        }
        if (root->right) {
            int right = helper(root->right);
            if (right > 0) temp += right;
            res = max(res, right);
        }
        temp += root->val;
        if (temp > maxSum) maxSum = temp;
        
        res += root->val;
        return res;
    }
};